File: #0576.out-of-boundary-paths.cpp
class Solution {
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {

        long path[m][n][maxMove + 1]; 
        memset(path, -1, sizeof(path));
        function<long(int, int, int)> fnPath = [&](int i, int j, int mv) {
            if (i < 0 || i >= m || j < 0 || j >= n) {
                return 1l;
            }

            if (mv == 0) {
                return 0l;
            }

            if (path[i][j][mv] < 0) {
                path[i][j][mv] = (
                    fnPath(i - 1, j, mv - 1) +
                    fnPath(i, j - 1, mv - 1) +
                    fnPath(i, j + 1, mv - 1) +
                    fnPath(i + 1, j, mv - 1)) % 1'000'000'007; 
            }

            return path[i][j][mv]; 
        }; 
        
        return fnPath(startRow, startColumn, maxMove); 
    }
};

File: #0576.out-of-boundary-paths.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number} maxMove
 * @param {number} startRow
 * @param {number} startColumn
 * @return {number}
 */
var findPaths = function(m, n, maxMove, startRow, startColumn) {
    let dp = {};
    return dfs(startRow, startColumn, maxMove);

    function dfs(startRow, startColumn, maxMove) {
        if (startRow < 0
            || startRow == m
            || startColumn < 0
            || startColumn == n) {
            return 1;
        }

        if (maxMove == 0) {
            return 0;
        }
        const key = `${startRow}#${startColumn}#${maxMove}`;
        if (dp[key] != undefined) {
            return dp[key];
        }
        
        let result = 0;
        result += dfs(startRow - 1, startColumn, maxMove-1);
        result += dfs(startRow + 1, startColumn, maxMove-1);
        result += dfs(startRow, startColumn-1, maxMove-1);
        result += dfs(startRow, startColumn+1, maxMove-1);
        result %= 1000000007;
        dp[key] = result;
        return result;
    };
};
